io.fairyproject.modules/platform.mc/module.hologram/src/main/java/io/fairyproject/mc/hologram/HologramImpl.java

```diff
@@ -0,0 +1,475 @@
+    private final Set<Consumer<MCPlayer>> attackHandlers;
+    private final Set<Consumer<MCPlayer>> interactHandlers;
+    private final List<HologramLine> lines;
+    private final List<HologramEntity> entities;
+
+    public HologramImpl(@NotNull Pos pos) {
+        this.world = pos.getMCWorld();
+        this.pos = pos;
+        this.autoViewable = true;
+        this.verticalSpacing = 0.25;
+        this.viewDistance = 4;
+        this.lines = new ArrayList<>();
+        this.entities = new ArrayList<>();
+        this.viewers = new ConcurrentHashMap<>();
+        this.attackHandlers = ConcurrentHashMap.newKeySet();
+        this.interactHandlers = ConcurrentHashMap.newKeySet();
+    }
+
+    @Override
+    public Hologram withAutoViewable(boolean autoViewable) {
+        this.autoViewable = autoViewable;
+        return this;
+    }
+
+    @Override
+    public Hologram withViewDistance(int viewDistance) {
+        this.viewDistance = viewDistance;
+        return this;
+    }
+
+    @Override
+    public Hologram withAttackHandler(Consumer<MCPlayer> attackHandler) {
+        this.attackHandlers.add(attackHandler);
+        return this;
+    }
+
+    @Override
+    public Hologram withInteractHandler(Consumer<MCPlayer> interactHandler) {
+        this.interactHandlers.add(interactHandler);
+        return this;
+    }
+
+    @Override
+    public Hologram withLines(@NotNull List<HologramLine> lines) {
+        synchronized (this) {
+            this.lines.clear();
+            this.lines.addAll(lines);
+        }
+        this.updateEntities();
+        return this;
+    }
+
+    @Override
+    public Hologram withLine(@NotNull HologramLine line) {
+        synchronized (this) {
+            this.lines.add(line);
+        }
+        this.updateEntities();
+        return this;
+    }
+
+    @Override
+    public Hologram withLine(int index, @NotNull HologramLine line) {
+        synchronized (this) {
+            this.lines.set(index, line);
+        }
+        this.updateEntities();
+        return this;
+    }
+
+    @Override
+    public Hologram withPos(@NotNull Pos pos) {
+        if (this.pos.getMCWorld() != this.world)
+            throw new IllegalArgumentException("hologram doesn't support cross world teleportation.");
+        this.pos = pos;
+        this.viewers.keySet().forEach(this::update);
+        return this;
+    }
+
+    @Override
+    public Hologram withVerticalSpacing(double verticalSpacing) {
+        this.verticalSpacing = verticalSpacing;
+        return this;
+    }
+
+    @Override
+    public Hologram withAttached(@Nullable MCEntity entity) {
+        this.attached = entity;
+        this.viewers.keySet().forEach(this::update);
+        return this;
+    }
+
+    @Override
+    public void removeLine(int index) {
+        this.lines.remove(index);
+    }
+
+    @Override
+    public void clear() {
+        this.lines.clear();
+    }
+
+    @Override
+    public Hologram spawn() {
```

Comment:
If im right, you never verify that the "pos" is not null which is gonna throw NPE in the future

Reply:
fixed